src/Listener/DefaultSender.php

```diff
@@ -92,18 +93,27 @@ public function onSend(SendBatchEvent $e) {
         /** @var \PEAR_Error $result */
         // CRM-9191
         $message = $result->getMessage();
+        // SMTP response code is buried in the message.
+        $code = preg_match('/ \(code: (.+), response: /', $message, $matches) ? $matches[1] : '';
         if (
           strpos($message, 'Failed to write to socket') !== FALSE ||
-          strpos($message, 'Failed to set sender') !== FALSE
+          ((
+            strpos($message, 'Failed to set sender') !== FALSE ||
+            strpos($message, 'Failed to add recipient') !== FALSE ||
+            strpos($message, 'Failed to send data') !== FALSE
+          // Register 5xx SMTP response code (permanent failure) as bounce.
+          ) && substr($code, 0, 1) !== '5')
         ) {
           // lets log this message and code
           $code = $result->getCode();
           \CRM_Core_Error::debug_log_message("SMTP Socket Error or failed to set sender error. Message: $message, Code: $code");
 
           // these are socket write errors which most likely means smtp connection errors
-          // lets skip them
+          // lets skip them and reconnect.
           $smtpConnectionErrors++;
           if ($smtpConnectionErrors <= 5) {
+            $mailer->disconnect();
+            $retryBatch = TRUE;
```

Comment:
Trying to verbalize what happens here... I'm not certain that these interpretations are correct (it's a long function with a lot of branching), but maybe verbalizing will ensure we're on the same page...

1. On the first temporary/retriable SMTP failure (and possibly again for a total of five temporary/retriable SMTP failures):
    1. We decline to write any status update about this particular message (i.e. we don't do anything with `$deliveredParams`, `$targetParams`, or `writeToDB()`). This implicitly means that it hasn't been delivered. Future runs of `Job.process_mailing` should retry this particular message.
    2. We know that there's unfinished work. After the DefaultSender completes this pass through the list of tasks, we should say `$e->setCompleted(FALSE)` (regardless of anything else). So we set`$retryBatch=TRUE`. 
        1. This does *not* mean that the entire batch will be retried. It just means that more work remains to be done. When a future worker grabs a batch of tasks, it should include anything that we didn't address.
    3. We disconnect from the SMTP server. This SMTP server is feeling cranky -- let's hope for better luck next time. When the next loop runs and uses`$mailer` again, it should make a new connection.
2. If we get 5+ temporary SMTP errors, give up temporarily. Wait for a future run of `Job.process_mailing`.